COMPACT_DENYLIST_FORMAT.md

```diff
@@ -0,0 +1,458 @@
+
+The BlockService should:
+
+- Convert the CID to `CIDV1BASE32/` (keeping the CID codec and adding a slash at the end) and hash it with the hashing functions used in the denylist. Match against declared double-hashes.
+
+- Convert the CID to b58-encoded-multihash (that is CIDv0) and hash the CID string.
+
+##### `/mime/MIMETYPE` `/mime/*`
+
+Blocks content detected to be of the given type. `/mime/*` blocks all the mimetypes and is meant to work with allow rules (all mimetypes blocked except specific ones).
+
+Blocking layer recommendation: Unixfs
+
+Our recommendation is that /mime/ rules automatically set IPFS clients into a
+"unixfs only" mode where only unixfs (+raw blocks) are allowed at the
+BlockService layer, and content type is checked at the Unixfs layer, as the
+blocks get assembled into an actual files. That should cover gateway usage.
+
+#### Allow (or negated) rules
+
+Block items can be prepended by `+`, that items matching the rule are to be allowed
+rather than blocked.
+
+This can be used to undo existing rules, but also to add concrete exceptions to wider rules. Order matters, and Allow rules must come AFTER other existing rules.
+
+Implementations should parse rules in general, and match them in inverse order
+as they appear in the denylist, so an explicit Allow rule will be evaluated
+before previously defined Deny rules, and can return non-blocked status for an
+item before further processing.
+
+Examples:
+
+```
+/ipfs/QmecDgNqCRirkc3Cjz9eoRBNwXGckJ9WvTdmY16HP88768/photo*
++/ipfs/QmecDgNqCRirkc3Cjz9eoRBNwXGckJ9WvTdmY16HP88768/photo123.jpg
+/mime/*
++/mime/text/plain
++/ipns/my.domain
+/ipns/my.domain
+```
+
+In this example, `/ipns/my.domain` stays blocked because the deny rule happens
+after the allow one.
+
+#### Hint list
+
+A hint list is an optional space-separated list of hints associated with specific block items in the form:
+
+```
+<block_item> hintA:v1 hintB:v2 hintC:v3
+```
+
+Block items and hints are separated by one or more consecutive instances of
+the "space" character.
+
+### Test fixtures
+
+Denylist parsing and correct behaviour can be tested using the
+[test.deny](https://github.com/ipfs-shipyard/nopfs/blob/master/tester/test.deny)
+denylist, which provides example rules and describes the expected behaviour in
+detail.
+
+In particular, a [Blocker implementation validator](https://github.com/ipfs-shipyard/nopfs/tree/master/tester) is provided in Go, and can be adapted to other languages if needed.
+
+### Security
+
+This proposal takes into account security:
+
+- Denylist headers and line-length limits are well specified to avoid malformed lists to cause things like large memory usage while parsing.
+- Supported type of blocks have been though out to avoid amplified consumption of resources or side effects (i.e. downloading of additional dag-blocks) during the implementation.
```

Comment:
```suggestion
- Supported type of blocks have been thought out to avoid amplified consumption of resources or side effects (i.e. downloading of additional dag-blocks) during the implementation.
```